[Kernel] Add per-tensor and per-token AZP epilogues

ProExpertProg opened the pull request:
Added two new epilogues that support `int32` azp, one for per-token and per-tensor each. Also added a unit test for folding the AZP correction term into the bias, which is disabled for now due to precision issues.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

ProExpertProg committed:
working version of folding bias into azp

ProExpertProg committed:
per-token quantization

ProExpertProg committed:
fixed the quantized zp case

ProExpertProg committed:
removed TODO

ProExpertProg committed:
bias is out dtype now

ProExpertProg committed:
naive_mm extracted

ProExpertProg committed:
Added comments to BiasEpilogue

ProExpertProg committed:
added ScaledEpilogueBiasAzp with per-token azp

DarkLight1337 commented:
To speed up the CI queue, I've cancelled the distributed tests for the latest CI run in this PR since they won't pass anyway until #5905 has been merged. Now that it has been merged, please merge `main` into your branch so that the CI can pass once again.

ProExpertProg commented:
No problem, thanks for letting me know! This is a draft so there's no rush, will rebase at some point when I'm back from vacation.

ProExpertProg committed:
per-token azp epilogue for other sm types (not just sm80)

ProExpertProg committed:
refactored w8a8 benchmarks and added bias and azp cases

ProExpertProg committed:
format

ProExpertProg committed:
w8a8 for fp8

tlrmchlsmth commented on tests/kernels/test_cutlass.py:
```diff
@@ -221,6 +225,109 @@ def test_cutlass_int8_gemm_m_sweep(per_act_token: bool, per_out_ch: bool,
+    azp_a = azp_aq_i8.to(dtype=torch.float32) * scale_a  # correct for rounding
+
+    a_dq = scale_a * (aq_i32 + azp_aq_i8).to(dtype=torch.float32)
+    torch.allclose(a_dq, scale_a * aq_f32 + azp_a)
```
ditto, assert

tlrmchlsmth reviewed (commented):
It looks like the epilogues are performing the computation:
`scale_a * scale_b * Acc + (scale_a * a_zp) * (scale_b * azp_adj)`

We will have less rounding error if we instead do:
`scale_a * scale_b * (Acc + a_zp * azp_adj)`

Do you see any issues with implementing it this way?

ProExpertProg committed:
PR comments, tests more robust

ProExpertProg committed:
using float for compute type in per-token azp

ProExpertProg committed:
refactored epilogues by extracting utilities

ProExpertProg committed:
per-token azp done in integers before scaling

ProExpertProg committed:
renamed epilogue, fixed comments

ProExpertProg committed:
Added per-tensor azp epilogue (int32) and added a test

ProExpertProg committed:
Fixed assert for de-quantizing in tests

ProExpertProg committed:
format

ProExpertProg committed:
added bias to azp tests

ProExpertProg committed:
visual load reduced, azp i8/i32 TODO, azp-fold test skipped

ProExpertProg committed:
merged per-token and per-tensor azp tests into one function to reduce duplication

ProExpertProg committed:
AZP epilogue fixed for sm75 and sm89

ProExpertProg committed:
AZP epilogues for c3x

ProExpertProg committed:
Expanded benchmarking cases

ProExpertProg committed:
Cache zero bias tensor

ProExpertProg committed:
nullptr bias working

ProExpertProg committed:
added nullptr support to ColOrScalar

ProExpertProg committed:
If nullptr enabled, no scalar load

ProExpertProg committed:
Tightened tolerance for AZP tests

ProExpertProg committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into luka/aq-azp-test

# Conflicts:
#	benchmarks/cutlass_benchmarks/w8a8_benchmarks.py
#	csrc/quantization/cutlass_w8a8/scaled_mm_entry.cu

ProExpertProg committed:
Cleanup: azp tensors can be 2d, removed BiasCache, comments

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -124,24 +159,31 @@ struct ScaledEpilogue
 };
 
+/*
+ * This epilogue performs the same operation as ScaledEpilogue, but adds a bias.
+ * This bias can also be used in the per-tensor azp case, where the activation
+ * zero point (azp) is folded into the bias.
```
I think we should clarify this comment: We don't we fold the activation zero point into the bias -- we compute a correction term that accounts for the activation zero point, and fold the AZP correction term into the bias

tlrmchlsmth commented on csrc/torch_bindings.cpp:
```diff
@@ -142,13 +142,22 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
       "                  Tensor b_scales, Tensor? bias) -> ()");
   ops.impl("cutlass_scaled_mm", torch::kCUDA, &cutlass_scaled_mm);
 
+  // CUTLASS w8a8 GEMM, supporting asymmetric per-tensor or per-row/column
+  // quantization.
```
Could you add a couple of sentences describing what `azp` and `azp_adj` are? I think we should try to give the reader enough information to know how to compute them offline.

ProExpertProg committed:
comments

cyang49 reviewed (commented):
This is a complex PR. I won't pretend that I understand all of it. I added some comments, but you might want a person who's more experienced in cutlass to approve this

ProExpertProg commented on csrc/quantization/cutlass_w8a8/scaled_mm_entry.cu:
```diff
@@ -105,3 +141,58 @@ void cutlass_scaled_mm(torch::Tensor& c, torch::Tensor const& a,
+                           torch::Tensor const& b_scales,
+                           c10::optional<torch::Tensor> const& bias,
+                           c10::optional<torch::Tensor> const& azp,
+                           torch::Tensor const& azp_adj) {
```
Yeah I think that's better. I wanted to make the change but forgot

ProExpertProg commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -554,4 +717,27 @@ void cutlass_scaled_mm_sm90(torch::Tensor& c, torch::Tensor const& a,
+    return cutlass_scaled_mm_sm90_epilogue<ScaledEpilogueBiasAzpToken>(
+        out, a, b, a_scales, b_scales, bias, *azp, azp_adj);
+  } else {
+    return cutlass_scaled_mm_sm90_epilogue<ScaledEpilogueBiasAzp>(
```
That's correct! There's a balance between performance and code size with additional epilogues as each generates a new kernel. So we have a separate bias epilogue in the symmetric case for better performance but not in the asymmetric case for smaller code size (and slightly worse performance).

ProExpertProg commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -154,31 +197,151 @@ struct ScaledEpilogueBias
+  using Azp = typename SUPER::template ColLoad<int32_t>;
+
+  // This is the AZP adjustment term, J @ B, shape (1,n)
+  using AzpAdj = typename SUPER::template RowLoad<int32_t>;
```
Yes, this term is always needed. Azp is sometimes needed (here), and when not needed, included in the `azp_adj` term, which is included in `AzpWithAdj` in the per-tensor epilogue.

tlrmchlsmth commented on csrc/torch_bindings.cpp:
```diff
@@ -142,13 +142,22 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, ops) {
   ops.impl("cutlass_scaled_mm", torch::kCUDA, &cutlass_scaled_mm);
 
+  // CUTLASS w8a8 GEMM, supporting asymmetric per-tensor or per-row/column
+  // quantization.
```
@cyang49 there are some details on these terms in our int8 activation quantization RFC https://github.com/vllm-project/vllm/issues/3975. Look at the Zero Point Correction terms section where it talks about handling $z_A J_A \widehat B$ for per-token quantization.

cyang49 replied:
Thanks for the explanation

tlrmchlsmth reviewed (commented):
@ProExpertProg I think we need to include an in-depth doc somewhere on how to handle zero points in general using the correction terms computed offline, and specifically the $z_A J_A \widehat B$ term for the per-token case. 

What do folks think about adding a markdown file (maybe in the `cutlass_w8a8` directory), so we can render LaTeX equations describing what the epilogues are for?

ProExpertProg commented on vllm/_custom_ops.py:
```diff
@@ -241,6 +243,29 @@ def cutlass_scaled_mm(a: torch.Tensor,
+                          out_dtype: Type[torch.dtype],
+                          azp: Optional[torch.Tensor],
+                          azp_adj: torch.Tensor,
+                          bias: Optional[torch.Tensor] = None) -> torch.Tensor:
```
No, good catch

ProExpertProg committed:
Added epilogues doc

ProExpertProg committed:
Math fixes

ProExpertProg committed:
Math fixes

ProExpertProg committed:
Reordered parameters for consistency

ProExpertProg committed:
Changed zero-points to be negative

ProExpertProg committed:
Arg order benchmarks

ProExpertProg commented:
@cyang49 I've addressed all of your comments, could you take a final look? I also added the `Epilogues.md` doc with extended descriptions and inverted the sign of azp to be consistent with the RFC.

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/broadcast_load_epilogue_c2x.hpp:
In this file, you've rearranged code a bit more than necessary. Don't get me wrong: it's a nice refactor, but could you instead try to isolate your changes as much as possible so that it's easier to keep this file in sync with the CUTLASS upstream?

ProExpertProg committed:
Unrefactored c2x visitor to stay more similar to original

ProExpertProg committed:
fixed test case azp_adj dtype

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:
```diff
@@ -77,19 +77,56 @@ struct enable_sm89_to_sm90 : Kernel {
+  static auto args_from_tensor(torch::Tensor const& tensor) {
+    using Arguments = typename Descriptor::Arguments;
+    auto* data_ptr = static_cast<T*>(tensor.data_ptr());
+    if constexpr (std::is_same_v<Descriptor, ColOrScalarLoad<T>> ||
+                  std::is_same_v<Descriptor, RowOrScalarLoad<T>>) {
+      return Arguments{data_ptr, tensor.numel() != 1};
+    } else {
+      return Arguments{data_ptr};
+    }
```
What happens if `Descriptor` is a `ColOrScalarLoad<T, true>`? Seems like I should still be able to pass in a non-optional tensor in that case, but it looks like we would end up with an illegal memory access in that case

ProExpertProg replied:
Added an assert to prevent that case, it shouldn't be necessary: if you know there's a tensor, there's no point in using the nullptr functionality. And if you really need it you can convert to optional before passing

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:
```diff
@@ -77,19 +77,56 @@ struct enable_sm89_to_sm90 : Kernel {
-  using ScaleB = cutlass::epilogue::threadblock::VisitorRowOrScalarBroadcast<
-      OutputTileThreadMap, float, Stride<Int<0>, Int<1>, Int<0>>>;
+  template <typename Descriptor, typename T>
+  static auto args_from_tensor(c10::optional<torch::Tensor> const& tensor) {
```
Could you add a brief comment describing the difference between the `args_from_tensor` functions? It'd be easier to read if there's a comment pointing out the `c10::optional` and the case each fn handles

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/broadcast_load_epilogue_c2x.hpp:
```diff
@@ -213,7 +239,8 @@ struct VisitorRowOrScalarBroadcast {
   class ThreadMap,
   class Element,
-  class StrideMNL = Stride<_1,_0,_0>
+  class StrideMNL = Stride<_1,_0,_0>,
+  bool EnableNullptr = false
```
Do we ever exercise the `EnableNullptr = true` case for `VisitorColOrScalarBroadcast`? 

In general, it's best to not write code that's untested and exercised, as untested code should be assumed to be buggy. I suggest keeping the interface as-is but adding a static assert with a comment that the case is unimplemented.

ProExpertProg replied:
Should I remove the changes?

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
````diff
@@ -0,0 +1,119 @@
+
+If $` \widehat X `$ is the quantized $` X `$, our matrices become the following
+
+```math
+A = s_a (\widehat A - J_a z_a)
+B = s_b \widehat B
+D = A B + C
+D = s_a s_b \widehat D + C
+```
````
could you define J_a, z_a, s_a, and s_b?

tlrmchlsmth replied:
Also, these aren't multi-line equations, at least when viewing them as a "rich diff" in this PR. Could you check your formatting here? Ditto for the rest of the equations.

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
```diff
@@ -0,0 +1,119 @@
+- `scale_b` is the scale for weights, can be per-tensor (scalar) or per-channel (row-vector).
+- `azp_with_adj` is the precomputed zero-point term ($` z_a J_a \widehat B `$), is per-channel (row-vector).
+- `bias` is the bias, is always per-channel (row-vector).
+
```
Could you add the following call-to-action to make it really clear what users of these kernels need to do in order to be able to use them?
**In order to use these kernels efficiently, users must precompute the `azp_with_adj` term offline in order to account for the activation zero point.**

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
````diff
@@ -0,0 +1,119 @@
+The epilogue performs the following computation (where `Dq` is the raw quantized output of the GEMM):
+```
+out = scale_a * scale_b * (Dq - azp_adj * azp) + bias
+```
````
Similar, add a call-to-action here for precomputing the `azp_adj` term

tlrmchlsmth commented:
@ProExpertProg the PR looks good, but could use some clarification in a couple of spots. Sidenote: I was hoping that the new CUTLASS 3.5.1 would handle the 2.x `null_default` case but unfortunately no

ProExpertProg committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into luka/aq-azp-test

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/broadcast_load_epilogue_c2x.hpp:
```diff
@@ -283,7 +311,26 @@ struct VisitorColOrScalarBroadcast {
+      if (EnableNullptr || params_ptr->col_broadcast) {
         // In this case we are loading from a column vector and broadcasting
         copy_if(pred, tC_gCol, tC_rCol);
       } else {
```
you created a `fill_dst` utility thats a copy of this `else` statement, why not use it there too?

ProExpertProg replied:
Tyler told me to make as few changes as possible to existing codepaths. We're also hoping to revert this soon once my CUTLASS PR lands

ProExpertProg commented on csrc/quantization/cutlass_w8a8/broadcast_load_epilogue_c2x.hpp:
```diff
@@ -134,7 +136,31 @@ struct VisitorRowOrScalarBroadcast {
+        }
+      }
+
+      // if nullptr is enabled but not present, always row broadcast
+      if (EnableNullptr || params_ptr->row_broadcast) {
```
that means the actual tensor passed is not a `nullptr`

ProExpertProg replied:
Is there a better way to phrase that?

LucasWilkinson replied:
I agree with @tlrmchlsmth (https://github.com/vllm-project/vllm/pull/5941/files#r1697288192), this logic is extremely confusing, my understanding is the logic is as follows:

```
should_scalar_broadcast = EnableNullptr ? 
     (params_ptr->ptr_row == nullptr) :
     (!params_ptr->row_broadcast);

if (should_scalar_broadcast):
     scalar_val = EnableNullptr ? 
         params_ptr->null_default : // (always 0 as far as I can tell)
         *params_ptr->ptr_row;
     scalar_broadcast(scalar_val)
else:
     row_broadcast(params_ptr->ptr_row)
```

but currently its written as:
```
if (EnableNullptr):
     // params_ptr->row_broadcast becomes irrelevant 
     if (params_ptr->ptr_row == nullptr):
        scalar_broadcast(params_ptr->null_default) // (always 0 as far as I can tell)
        return 

// params_ptr->ptr_row == nullptr is now invalid
if (EnableNullptr || params_ptr->row_broadcast):
    row_broadcast(params_ptr->ptr_row)
else:
    scalar_broadcast(*params_ptr->ptr_row)
```

is this correct? I think the current logic obfuscates the impact `EnableNullptr` flag has on the scalar broadcast case logic (i.e. changes the condition in which we scalar broadcast and the source of the scalar for broadcasting)

I think a cleaner versions would be to just say `ptr_row` is always the scalar source if we want a non-zero/dynamic scalar broadcast, and when `EnableNullptr` is true then a null `ptr_row` always becomes a noop, i.e.:

```
if (EnableNullptr && (params_ptr->ptr_row == nullptr)):
     scalar_broadcast(0) // noop
elif (params_ptr->row_broadcast):
     row_broadcast(params_ptr->ptr_row)
else:
     scalar_broadcast(*params_ptr->ptr_row)
     
```

I think this would dramatically simplify the logic here, although I understand arguments for keeping `null_default` around for consistency with upstreams sake. Whatever the case I think we should make an effort to clarify the logic here, either by adding comments/docs or restructuring some of the conditions

ProExpertProg replied:
Those are all valid points. But performance is the reason this is structured the way it is. `EnableNullptr` is known at compile time, so I didn't want the code to also branch on `row_broadcast` after already checking for `nullptr`.

I added a comment, I hope that clears things up. I also wanted to keep the changes minimal. So this monstrosity is the result.

I could also just make YET ANOTHER copy of the row-broadcast visitor that supports the nullptr. That would be cleaner but have a higher overall change footprint.

LucasWilkinson replied:
> But performance is the reason this is structured the way it is.
im not sure this a huge concern here given you'd get a "performance" (not convinced any of these things we are discussing are that consequential) win in the noop case the 0 broadcasted value is now known at compile time (can just 0 registers instead of loading `params_ptr->null_default`)
> I also wanted to keep the changes minimal.
Fair enough, I didn't realize the indentation was for this code to be temporary, that definitely changes how I view it. We should a comment mentioning the code is temporary then.

the comment is better, thank you!

tlrmchlsmth replied:
I think the fact that `params_ptr->row_broadcast` becomes irrelevant if EnableNullptr is my biggest problem with this code. It's not clear from the interfaces to this class, even with the latest added comments on the `Arguments` struct. I'm also not confident that any of our changes will end up making their way into upstream CUTLASS, so would prefer a less confusion solution here. 

I think @LucasWilkinson suggestion would be a big improvement here

ProExpertProg replied:
Went with `RowOrZeroBroadcast`

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/broadcast_load_epilogue_c2x.hpp:
```diff
@@ -134,7 +136,31 @@ struct VisitorRowOrScalarBroadcast {
 
-      if (params_ptr->row_broadcast) {
+      // Fill dst_v with the scalar value
+      auto fill_dst = [&](Element const& value) {
```
you created this `fill_dst` utility thats a copy of this `else` statement at line 171, why not use it there too?

LucasWilkinson commented on csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cuh:
```diff
@@ -166,30 +207,163 @@ struct ScaledEpilogueBias
+ * O(m*n) space, and this way it only requires O(m+n) space.
+ * The azp term is a 1D tensor of shape (m,1), and represents the unscaled zero
+ * point for each row of A.
+ * The azp_adj term is a 1D tensor of shape (1,n), computed as J @ B.
```
what is `J @ B` be here, why not just use names consistent with the comment? i.e. `azp @ azp_adj`?

LucasWilkinson commented on tests/kernels/test_cutlass.py:
```diff
@@ -221,6 +224,135 @@ def test_cutlass_int8_gemm_m_sweep(per_act_token: bool, per_out_ch: bool,
                                      use_bias)
 
 
+def naive_mm(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
```
doesn't look like this is used? or am I missing something?

ProExpertProg replied:
Yep I fixed it, good catch

LucasWilkinson commented on vllm/_custom_ops.py:
```diff
@@ -560,7 +584,7 @@ def dispatch_bgmv_low_level(
         and v.__code__.co_filename == __file__ \
         and any(arg is torch.Tensor or arg == "torch.Tensor"
-                   for arg in v.__annotations__.values()):
+                for arg in v.__annotations__.values()):
```
why the format change?

ProExpertProg replied:
I don't know I didn't touch it 😮

ProExpertProg commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -154,31 +197,160 @@ struct ScaledEpilogueBias
-  using Bias = cutlass::epilogue::fusion::Sm90RowBroadcast<
-      BiasDescriptor::Stages, typename EpilogueDescriptor::TileShape, ElementD,
-      Stride<Int<0>, Int<1>, Int<0>>, 128 / sizeof_bits_v<ElementD>, false>;
+  // Compute (accum - azp_adj), resulting in a float
```
`accum` and `azp_adj` are both `int32_t` so I don't want people to confuse the cast with the type. But I'll make it clearer

ProExpertProg committed:
PR comments

ProExpertProg committed:
PR comments: md

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
````diff
@@ -0,0 +1,136 @@
+Expanding further, we can calculate $` \widehat D `$ as follows:
+
+```math
+A B = s_a ( \widehat A - J_a z_a ) s_b \widehat B = s_a s_b \left( \widehat A \widehat B - J_a z_a \widehat B \right)
````
Please break up these equation

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
````diff
@@ -0,0 +1,136 @@
+\widehat D = \widehat A \widehat B
+```
+```math
+D = s_a s_b \widehat D = s_a s_b \widehat A \widehat B
````
ditto

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
````diff
@@ -0,0 +1,136 @@
+\widehat D = \widehat A \widehat B
+```
+```math
+D = s_a s_b \widehat D + C = s_a s_b \widehat A \widehat B + C
````
ditto

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/Epilogues.md:
````diff
@@ -0,0 +1,136 @@
+\widehat D = \widehat A \widehat B - z_a J_a \widehat B
+```
+```math
+D = s_a s_b \widehat D + C = s_a s_b \left( \widehat A \widehat B - z_a J_a \widehat B \right) + C
````
ditto

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -60,25 +60,66 @@ struct enable_sm90_or_later : Kernel {
+  // This utility function constructs the arguments for the load descriptors
+  // from a tensor. It can handle both row and column, as well as row/column or
+  // scalar cases.
+  template <typename Descriptor, typename T>
+  static auto args_from_tensor(torch::Tensor const& tensor) {
+    using Arguments = typename Descriptor::Arguments;
+    auto* data_ptr = static_cast<T*>(tensor.data_ptr());
+    if constexpr (std::is_same_v<Descriptor, ColOrScalarLoad<T>> ||
+                  std::is_same_v<Descriptor, RowOrScalarLoad<T>>) {
+      return Arguments{data_ptr, tensor.numel() != 1};
+    } else {
+      return Arguments{data_ptr};
+    }
+  }
```
Please add static_asserts like you did for the the 2.x version

ProExpertProg commented on csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cuh:
```diff
@@ -75,19 +75,56 @@ struct enable_sm89_to_sm90 : Kernel {
+      return Arguments{data_ptr, tensor.numel() != 1};
+    } else {
+      return Arguments{data_ptr};
+    }
```
I added an assert, did you refresh the files?

ProExpertProg committed:
revert visitor

ProExpertProg committed:
added roworzero visitor, c2x uses it for bias in AZP cases, also assert in c3x

ProExpertProg committed:
PR comments: equations

ProExpertProg committed:
format

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cu:
```diff
@@ -57,6 +57,29 @@ void cutlass_scaled_mm_sm75(torch::Tensor& out, torch::Tensor const& a,
+                                torch::Tensor const& azp_adj,
+                                c10::optional<torch::Tensor> const& azp,
+                                c10::optional<torch::Tensor> const& bias) {
+  TORCH_CHECK(a_scales.dtype() == torch::kFloat32);
+  TORCH_CHECK(b_scales.dtype() == torch::kFloat32);
+  TORCH_CHECK(!bias || bias->dtype() == out.dtype(),
+              "currently bias dtype must match output dtype ", out.dtype());
+  TORCH_CHECK(azp_adj.dtype() == torch::kInt32);
+
+  if (azp) {
+    TORCH_CHECK(azp->dtype() == torch::kInt32);
```
It looks like these TORCH_CHECKs always the same. If so, could you move them into `cutlass_scaled_mm_azp` instead of having them be duplicated in each arch-specific function?

ProExpertProg replied:
Should I do that for scales as well?

ProExpertProg replied:
I'll do it in a cleanup PR later

tlrmchlsmth replied:
Yeah, scales as well would be good to factor out.

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c2x.cuh:
```diff
@@ -166,30 +214,163 @@ struct ScaledEpilogueBias
+};
+
+/*
+ * This epilogue directly supports per-token azp by materializing the correction
+ * term using an outer product. If the term was precomputed, it would require
+ * O(m*n) space, and this way it only requires O(m+n) space.
+ * The azp term is a 1D tensor of shape (m,1), and represents the unscaled zero
+ * point for each row of A.
+ * The azp_adj term is a 1D tensor of shape (1,n), computed as J @ B.
+ *
+ * This epilogue also supports bias, which remains per-channel.
+ */
```
Let's fixup this comment. I think it says the opposite of what you mean -- "materializing the correction term using an outer product" would mean that we compute the correction term with an outer product and have the whole M*N term in memory. I suggest something like the following: 
```suggestion
 * This epilogue supports per-token azp by computing and applying
 * the correction term using a rank-1 update. If the term were materialized, 
 * it would require O(m*n) space, and this way it only requires O(m+n) space.
 * The azp term is a 1D tensor of shape (m,1), and represents the unscaled zero
 * point for each row of A.
 * The azp_adj term is a 1D tensor of shape (1,n), computed as J @ B.
 *
 * This epilogue also supports bias, which remains per-channel.
 */
```

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -154,31 +205,160 @@ struct ScaledEpilogueBias
-    ScaleB_Args b_args{b_scales.data_ptr<float>(), b_scales.numel() != 1, {}};
-    Bias_Args bias_args{static_cast<ElementD*>(bias.data_ptr())};
+/*
+ * This epilogue directly supports per-token azp by materializing the correction
```
Same comment on materializing the correction term here.

tlrmchlsmth reviewed (commented):
Suggest small factoring of TORCH_CHECKs, and a couple of comment fixups. Looks good after that.

ProExpertProg committed:
AZP per-token comment

Co-authored-by: Tyler Michael Smith <tyler@neuralmagic.com>

ProExpertProg committed:
c3x azptoken comment

ProExpertProg committed:
moved TORCH_CHECKs to top function

ProExpertProg committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into luka/aq-azp-test

# Conflicts:
#	csrc/ops.h

tlrmchlsmth commented:
/ready

ProExpertProg committed:
fixed type annotation

ProExpertProg committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into luka/aq-azp-test

ProExpertProg committed:
w8a8_benchmarks fix

ProExpertProg committed:
Merge remote-tracking branch 'upstream/main' into luka/aq-azp-test

ProExpertProg committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into luka/aq-azp-test
